tools/heir-opt.cpp

```diff
@@ -331,94 +353,35 @@ void tosaToBooleanTfhePipeline(const std::string &yosysFilesPath,
-        // Secretize inputs
-        pm.addPass(createSecretize(SecretizeOptions{options.entryFunction}));
-
-        // TOSA to linalg
-        tosaToLinalg(pm);
-
-        // Bufferize
-        oneShotBufferize(pm);
-
-        // Affine
-        pm.addNestedPass<FuncOp>(createConvertLinalgToAffineLoopsPass());
-        pm.addNestedPass<FuncOp>(memref::createExpandStridedMetadataPass());
-        pm.addNestedPass<FuncOp>(affine::createAffineExpandIndexOpsPass());
-        pm.addNestedPass<FuncOp>(memref::createExpandOpsPass());
-        pm.addNestedPass<FuncOp>(affine::createSimplifyAffineStructuresPass());
-        pm.addPass(memref::createFoldMemRefAliasOpsPass());
-        pm.addPass(createExpandCopyPass());
-        pm.addNestedPass<FuncOp>(affine::createAffineLoopNormalizePass(true));
-        pm.addNestedPass<FuncOp>(affine::createLoopFusionPass(
-            0, 0, true, affine::FusionMode::Greedy));
-        pm.addPass(affine::createAffineScalarReplacementPass());
-        pm.addPass(createForwardStoreToLoad());
+                                const TosaToBooleanTfheOptions &options) {
+        tosaToBooleanCGGIPipelineBuilder(pm, options, yosysFilesPath, abcPath,
+                                         /*abcBooleanGates=*/true);
 
-        // Cleanup
-        pm.addPass(createMemrefGlobalReplacePass());
-        arith::ArithIntNarrowingOptions arithOps;
-        arithOps.bitwidthsSupported = {4, 8, 16};
-        pm.addPass(arith::createArithIntNarrowing(arithOps));
+        // Vectorize CGGI operations
+        pm.addPass(createStraightLineVectorizer(
+            StraightLineVectorizerOptions{.dialect = "cggi"}));
         pm.addPass(createCanonicalizerPass());
-        pm.addPass(createSCCPPass());
         pm.addPass(createCSEPass());
-        pm.addPass(createSymbolDCEPass());
+        pm.addPass(createSCCPPass());
 
-        pm.addPass(createWrapGeneric());
-        auto distributeOpts = secret::SecretDistributeGenericOptions{
-            .opsToDistribute = opsToDistribute};
-        pm.addPass(secret::createSecretDistributeGeneric(distributeOpts));
+        // CGGI to Tfhe-Rust exit dialect
+        pm.addPass(createCGGIToTfheRustBool());
+        // CSE must be run before canonicalizer, so that redundant ops are
+        // cleared before the canonicalizer hoists TfheRust ops.
+        pm.addPass(createCSEPass());
         pm.addPass(createCanonicalizerPass());
 
-        // Booleanize and Yosys Optimize
-        pm.addPass(createYosysOptimizer(yosysFilesPath, abcPath,
-                                        options.abcFast, options.unrollFactor,
-                                        Mode::Boolean));
-
-        // Lower combinational circuit to CGGI
-        pm.addPass(createForwardStoreToLoad());
-        pm.addPass(mlir::createCSEPass());
-        pm.addPass(secret::createSecretDistributeGeneric());
-        pm.addPass(comb::createCombToCGGI());
-        // Cleanup CombToCGGI
+        // Cleanup loads and stores
```

Comment:
Ah, thanks, I forgot to move the canonicalize/cse/sccp part, the rest is there

Reply:
I don't think it's worth factoring out just this cleanup in both passes, I can't put it in the same common tosaToCGGIPipeline since these occur after the individual scheme-to-library dialect passes, but I did add in the cleanup after tosaToCGGI.